File: backend/app/services/remna_client.py
import httpx
from typing import Optional, List, Dict, Any
from app.core.config import settings


class RemnaClient:
    """Client for interacting with Remna API"""
    
    def __init__(self):
        self.base_url = settings.REMNA_PANEL_URL.rstrip('/')
        self.api_token = settings.REMNA_API_TOKEN
        self.headers = {
            "Authorization": f"Bearer {self.api_token}",
            "Content-Type": "application/json"
        }
    
    async def _request(
        self,
        method: str,
        endpoint: str,
        data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Make HTTP request to Remna API"""
        url = f"{self.base_url}/api{endpoint}"
        
        print(f"🔍 Making {method} request to {url}")
        if data:
            print(f"🔍 Request data: {data}")
        if params:
            print(f"🔍 Request params: {params}")
        
        async with httpx.AsyncClient() as client:
            response = await client.request(
                method=method,
                url=url,
                headers=self.headers,
                json=data,
                params=params,
                timeout=30.0
            )
            
            print(f"🔍 Response status: {response.status_code}")
            print(f"🔍 Response text: {response.text}")
            
            response.raise_for_status()
            return response.json()
    
    # User Management
    async def get_users(
        self,
        offset: int = 0,
        limit: int = 100,
        username: Optional[str] = None,
        status: Optional[str] = None
    ) -> Dict[str, Any]:
        """Get list of users"""
        params = {
            "offset": offset,
            "limit": limit
        }
        if username:
            params["search"] = username  # Remna API uses 'search' parameter
        if status:
            params["status"] = status
            
        result = await self._request("GET", "/users", params=params)
        # Remna API returns data in 'response' field
        return result.get("response", result)
    
    async def get_user(self, username: str) -> Dict[str, Any]:
        """Get user by username"""
        print(f"🔍 Getting user: {username}")
        result = await self._request("GET", f"/users/{username}")
        print(f"🔍 Get user response: {result}")
        return result.get("response", result)
    
    async def create_user(self, user_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create new user"""
        print(f"🔍 Creating user with data: {user_data}")
        result = await self._request("POST", "/users", data=user_data)
        print(f"🔍 Create user response: {result}")
        return result.get("response", result)
    
    async def update_user(self, username: str, user_data: Dict[str, Any]) -> Dict[str, Any]:
        """Update user - uuid is passed in the body, not username"""
        # Get user by username to get UUID
        user_info = await self.get_user(username)
        uuid = user_info.get('uuid')
        
        if not uuid:
            raise ValueError(f"User {username} not found or UUID missing")
        
        # Add uuid to the data payload as per Remna API requirements
        user_data['uuid'] = uuid
        result = await self._request("PATCH", "/users", data=user_data)
        return result.get("response", result)
    
    async def delete_user(self, username: str) -> Dict[str, Any]:
        """Delete user"""
        return await self._request("DELETE", f"/users/{username}")
    
    async def reset_user_traffic(self, username: str) -> Dict[str, Any]:
        """Reset user traffic"""
        return await self._request("POST", f"/users/{username}/reset")
    
    async def revoke_subscription(self, username: str) -> Dict[str, Any]:
        """Revoke user subscription"""
        return await self._request("POST", f"/users/{username}/revoke")
    
    # System & Stats
    async def get_system_stats(self) -> Dict[str, Any]:
        """Get system statistics from /api/system/stats"""
        try:
            result = await self._request("GET", "/system/stats")
            return result
        except Exception:
            # Если эндпоинт не существует, возвращаем пустой словарь
            return {}
    
    async def get_core_stats(self) -> Dict[str, Any]:
        """Get core statistics"""
        try:
            return await self._request("GET", "/core")
        except Exception:
            return {}
    
    async def get_nodes(self) -> List[Dict[str, Any]]:
        """Get nodes information"""
        try:
            result = await self._request("GET", "/nodes")
            return result if isinstance(result, list) else []
        except Exception:
            return []
    
    async def get_squads(self) -> List[Dict[str, Any]]:
        """Get internal squads information from /api/internal-squads"""
        try:
            result = await self._request("GET", "/internal-squads")
            # Remna API возвращает данные в формате {"response": {"internalSquads": [...]}}
            if isinstance(result, dict) and "response" in result:
                response = result["response"]
                if isinstance(response, dict) and "internalSquads" in response:
                    return response["internalSquads"]
            return []
        except Exception:
            return []


remna_client = RemnaClient()

File: backend/remnaClient.ts
import axios, { AxiosInstance, AxiosRequestConfig } from 'axios';
import { config } from './config';
import { User, UserCreate, UserUpdate, InternalSquad, HwidDevice, HwidDevicesResponse } from './types';

export class RemnaClient {
  private client: AxiosInstance;
  
  constructor() {
    this.client = axios.create({
      baseURL: `${config.remnaPanelUrl.replace(/\/$/, '')}/api`,
      headers: {
        'Authorization': `Bearer ${config.remnaApiToken}`,
        'Content-Type': 'application/json'
      },
      timeout: 30000
    });
  }
  
  private async request<T>(config: AxiosRequestConfig): Promise<T> {
    console.log(`🔍 Making ${config.method?.toUpperCase()} request to ${config.url}`);
    if (config.data) {
      console.log(`🔍 Request data:`, config.data);
    }
    if (config.params) {
      console.log(`🔍 Request params:`, config.params);
    }
    
    try {
      const response = await this.client.request<T>(config);
      console.log(`🔍 Response status: ${response.status}`);
      console.log(`🔍 Response data:`, response.data);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        console.error(`❌ Request failed:`, error.response?.data || error.message);
        throw error;
      }
      throw error;
    }
  }
  
  // User Management
  async getUsers(params: {
    offset?: number;
    limit?: number;
    search?: string;
    status?: string;
  }): Promise<{ users: User[]; total?: number }> {
    const response = await this.request<{ response: { users: User[] } }>({
      method: 'GET',
      url: '/users',
      params: {
        offset: params.offset || 0,
        limit: params.limit || 100,
        ...(params.search && { search: params.search }),
        ...(params.status && { status: params.status })
      }
    });
    
    return {
      users: response.response?.users || (response as any).users || [],
      total: (response as any).total
    };
  }
  
  async getUser(username: string): Promise<User> {
    console.log(`🔍 Getting user: ${username}`);
    const response = await this.request<{ response: User }>({
      method: 'GET',
      url: `/users/${username}`
    });
    console.log(`🔍 Get user response:`, response);
    return response.response || (response as any);
  }
  
  async createUser(userData: UserCreate): Promise<User> {
    console.log(`🔍 Creating user with data:`, userData);
    const response = await this.request<{ response: User }>({
      method: 'POST',
      url: '/users',
      data: userData
    });
    console.log(`🔍 Create user response:`, response);
    return response.response || (response as any);
  }
  
  async updateUser(username: string, userData: UserUpdate): Promise<User> {
    // Get user to retrieve UUID
    const userInfo = await this.getUser(username);
    const uuid = userInfo.uuid;
    
    if (!uuid) {
      throw new Error(`User ${username} not found or UUID missing`);
    }
    
    // Add uuid to the data payload as per Remna API requirements
    const updateData = {
      ...userData,
      uuid
    };
    
    const response = await this.request<{ response: User }>({
      method: 'PATCH',
      url: '/users',
      data: updateData
    });
    
    return response.response || (response as any);
  }
  
  async deleteUser(username: string): Promise<any> {
    return await this.request({
      method: 'DELETE',
      url: `/users/${username}`
    });
  }
  
  async resetUserTraffic(username: string): Promise<any> {
    return await this.request({
      method: 'POST',
      url: `/users/${username}/reset`
    });
  }
  
  async revokeSubscription(username: string): Promise<any> {
    return await this.request({
      method: 'POST',
      url: `/users/${username}/revoke`
    });
  }
  
  // System & Stats
  async getSystemStats(): Promise<any> {
    try {
      return await this.request({
        method: 'GET',
        url: '/system/stats'
      });
    } catch (error) {
      return {};
    }
  }
  
  async getCoreStats(): Promise<any> {
    try {
      return await this.request({
        method: 'GET',
        url: '/core'
      });
    } catch (error) {
      return {};
    }
  }
  
  async getNodes(): Promise<any[]> {
    try {
      const result = await this.request<any>({
        method: 'GET',
        url: '/nodes'
      });
      return Array.isArray(result) ? result : [];
    } catch (error) {
      return [];
    }
  }
  
  async getSquads(): Promise<InternalSquad[]> {
    try {
      const result = await this.request<{ response: { internalSquads: InternalSquad[] } }>({
        method: 'GET',
        url: '/internal-squads'
      });
      
      if (result.response?.internalSquads) {
        return result.response.internalSquads;
      }
      return [];
    } catch (error) {
      return [];
    }
  }

  // HWID Devices Management
  async getUserDevices(userUuid: string): Promise<HwidDevice[]> {
    try {
      const result = await this.request<HwidDevicesResponse | { response: HwidDevicesResponse }>({
        method: 'GET',
        url: '/hwid/devices',
        params: {
          userUuid
        }
      });
      
      if ((result as any).response?.devices) {
        return (result as any).response.devices;
      }
      if ((result as any).devices) {
        return (result as any).devices;
      }
      return [];
    } catch (error) {
      console.error('Failed to get user devices:', error);
      return [];
    }
  }

  async deleteDevice(deviceUuid: string): Promise<any> {
    return await this.request({
      method: 'DELETE',
      url: `/hwid/devices/${deviceUuid}`
    });
  }
}

export const remnaClient = new RemnaClient();

